llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -1023,11 +1023,30 @@ static void computeKnownBitsFromOperator(const Operator *I,
     break;
   }
   case Instruction::Select: {
-    computeKnownBits(I->getOperand(2), Known, Depth + 1, Q);
-    computeKnownBits(I->getOperand(1), Known2, Depth + 1, Q);
-
+    auto ComputeForArm = [&](bool Arm) {
+      unsigned ArmIdx = 1 + static_cast<unsigned>(!Arm);
+      KnownBits Res(Known.getBitWidth());
+      computeKnownBits(I->getOperand(ArmIdx), Res, Depth + 1, Q);
+      // See what condition implies about the bits of the two select arms.
+      if (!Known.isConstant()) {
+        KnownBits KnownFromCond(Known.getBitWidth());
+        computeKnownBitsFromCond(I->getOperand(ArmIdx), I->getOperand(0),
+                                 KnownFromCond, Depth + 1, Q, !Arm);
```

Comment:
This is incorrect if the value can be undef.

Reply:
Changing, but why is that the case? 
Something like: https://alive2.llvm.org/ce/z/P42d4W has no complaints.
I would think there would be no difference between trying to calculate an undef arm directly vs w/ using the condition bits.

Reply:
You need to consider a case where the instructions are not actually removed, e.g. https://alive2.llvm.org/ce/z/Fjkabr.

Reply:
Thanks :)